### PR TITLE
[FLINK-29194] Add logging for ResourceListener

### PR DESCRIPTION
**Pull request by morhidi**
## What is the purpose of the change

Being able to tell what status updates and events are being sent via the ResourceListener interfaces. It allows easy tail/grep for related messages reported under a specific resource.

## Brief change log
- Added `Unified audit logging for StatusRecorder/EventRecorder`
- Added `description for ResourceLifecycleState`
- Minor fix in EventRecorded to avoid creating redundant `ctx` objects if multiple ResourceListeners are registered.
## Verifying this change

Manual verification:
```
2022-09-07 06:53:23,661 o.a.f.k.o.l.AuditUtils         [INFO ] [default.basic-example] >>> Event  | Info    | SPECCHANGED     | UPGRADE change(s) detected (FlinkDeploymentSpec[serviceAccount=flink] differs from FlinkDeploymentSpec[serviceAccount=flinkk]), starting reconciliation.
2022-09-07 06:53:23,680 o.a.f.k.o.l.AuditUtils         [INFO ] [default.basic-example] >>> Event  | Info    | SUSPENDED       | Suspending existing deployment.
2022-09-07 06:53:24,849 o.a.f.k.o.l.AuditUtils         [INFO ] [default.basic-example] >>> Status | Info    | UPGRADING       | The resource is being upgraded 
2022-09-07 06:53:24,868 o.a.f.k.o.l.AuditUtils         [INFO ] [default.basic-example] >>> Status | Info    | UPGRADING       | The resource is being upgraded 
2022-09-07 06:53:24,884 o.a.f.k.o.l.AuditUtils         [INFO ] [default.basic-example] >>> Event  | Info    | SUBMIT          | Starting deployment
2022-09-07 06:53:25,126 o.a.f.k.o.l.AuditUtils         [INFO ] [default.basic-example] >>> Status | Info    | DEPLOYED        | The resource is deployed/submitted to Kubernetes, but it’s not yet considered to be stable and might be rolled back in the future 
```
## Does this pull request potentially affect one of the following parts:
  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changes to the `CustomResourceDescriptors`: no
  - Core observer or reconciler logic that is regularly executed: no

## Documentation
  - Does this pull request introduce a new feature? no

### CODE REVIEW

File: flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/utils/StatusRecorder.java [hunk 1]
```diff
@@ -196,6 +197,7 @@ public Instant getTimestamp() {
                                     listener.onSessionJobStatusUpdate(ctx);
                                 }
                             });
+                    ListenerUtils.logContext(ctx);
```
Comment (gaborgsomogyi) [inline flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/utils/StatusRecorder.java:200]: Not sure why we need to add this to `ListenerUtils`. If @gyfora pointed this out in the jira then +1 to inline it here :)

Comment (morhidi) [reply flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/utils/StatusRecorder.java:200]: With this approach we have a nice and aligned table on these logs, when I grep it is easier to grep on ListenerUtils.
```
2022-09-06 13:19:47,920 o.a.f.k.o.l.ListenerUtils      [INFO ] [default.basic-example] >>> Status | Info    | UPGRADING       | The resource is being upgraded 
2022-09-06 13:19:48,030 o.a.f.k.o.l.ListenerUtils      [INFO ] [default.basic-example] >>> Event  | Info    | Submit          | Starting deployment
2022-09-06 13:19:48,672 o.a.f.k.o.l.ListenerUtils      [INFO ] [default.basic-example] >>> Status | Info    | DEPLOYED        | The resource is deployed/submitted to Kubernetes, but it’s not yet considered to be stable and might be rolled back in the future 
2022-09-06 13:19:52,700 o.a.f.k.o.l.ListenerUtils      [INFO ] [default.basic-example] >>> Status | Info    | DEPLOYED        | The resource is deployed/submitted to Kubernetes, but it’s not yet considered to be stable and might be rolled back in the future 
2022-09-06 13:20:02,857 o.a.f.k.o.l.ListenerUtils      [INFO ] [default.basic-example] >>> Event  | Info    | StatusChanged   | Job status changed from RECONCILING to RUNNING
2022-09-06 13:20:02,927 o.a.f.k.o.l.ListenerUtils      [INFO ] [default.basic-example] >>> Status | Info    | STABLE          | The resource deployment is considered to be stable and won’t be rolled back 
```

Comment (morhidi) [reply flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/utils/StatusRecorder.java:200]: We're not simply logging for debugging purposes, I added extra logic to reflect the content users would normally see

Comment (morhidi) [reply flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/utils/StatusRecorder.java:200]: This should tell a nice story to the user.

File: flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/listener/ListenerUtils.java [hunk 2]
```diff
@@ -134,4 +138,32 @@ private static Set<Tuple2<String, String>> findEnabledListeners(Configuration co
...
+                status.getLifecycleState(),
+                StringUtils.isEmpty(status.getError())
+                        ? status.getLifecycleState().getDescription()
+                        : status.getError());
+    }
+
+    private static String format(@NonNull Event event) {
+        return String.format(
+                ">>> Event  | %-7s | %-15s | %s",
+                event.getType().equals("Normal") ? "Info" : event.getType(),
```
[the remark this answers is not in the document]
Comment (morhidi) [reply flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/listener/ListenerUtils.java:165]: We have Info/Error/Warning there, should be fine

File: flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/utils/EventRecorder.java [hunk 3]
```diff
@@ -69,34 +70,35 @@ public static EventRecorder create(
...
+                                }
+
+                                @Override
+                                public KubernetesClient getKubernetesClient() {
+                                    return client;
+                                }
+                            };
+                    listeners.forEach(
+                            listener -> {
+                                if (resource instanceof FlinkDeployment) {
```
[the remark this answers is not in the document]
Comment (morhidi) [reply flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/utils/EventRecorder.java:93]: Don't want to touch this part in this PR

File: flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/utils/StatusRecorder.java (hunk 1 above)
Comment (gaborgsomogyi) [reply flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/utils/StatusRecorder.java:200]: I agree w/ you basically. If I understand correctly this is mainly for user and tracks history.
Can we rename `ListenerUtils` to something more meaningful?

Comment (morhidi) [reply flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/utils/StatusRecorder.java]: Moved the logging to a new `AuditUtils` class which makes more sense agree :)

**Review by gaborgsomogyi — APPROVED** [review]
LGTM.

**Review by mbalassi — APPROVED** [review]
LGTM, will merge tomorrow if there are no further modification requests.

File: flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/listener/AuditUtils.java [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+                StringUtils.isEmpty(status.getError())
+                        ? status.getLifecycleState().getDescription()
+                        : status.getError());
+    }
+
+    private static String format(@NonNull Event event) {
+        return String.format(
+                ">>> Event  | %-7s | %-15s | %s",
+                event.getType().equals("Normal") ? "Info" : event.getType(),
+                event.getReason(),
```
Comment (SteNicholas) [inline flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/listener/AuditUtils.java:57]:
```suggestion
                event.getReason().toUpperCase(),
```

Comment (SteNicholas) [inline flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/listener/AuditUtils.java:57]:
```suggestion
                event.getReason().toUpperCase(),
```

Comment (morhidi) [reply flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/listener/AuditUtils.java]: it feels prettier :) fixed

**Review by SteNicholas — APPROVED** [review]
LGTM.

**Pull request merged**